drafts/refactoring-legacy-code-with-tiny-types-in-java.md

```diff
@@ -0,0 +1,237 @@
+}
+```
+
+As you can see, we use a `sealed` interface introduced with Java 17.
+[Sealed interfaces](https://openjdk.org/jeps/409) complement Java's enhanced [pattern matching](https://openjdk.org/jeps/441) by providing exhaustive case handling that the compiler enforces.
+This discussion naturally leads us to another topic I'd like to cover: [Data Oriented Programming in Java](https://www.infoq.com/articles/data-oriented-programming-java/), as explained in an InfoQ article by Brian Goetz.
+
+> _Data-oriented programming_ encourages us to model data as (immutable) data, and keep the code that embodies the business logic of how we act on that data separately.
+Records, sealed classes, and pattern matching are designed to work together to support data-oriented programming.
+Using the combination of records and sealed classes also makes it easier to _make illegal states unrepresentable_, further improving safety and maintainability.
+
+In the spirit of making illegal states unrepresentable, modeling our purchasable items from the shop begins here.
+For example, *"Sulfuras"* has a quality of `80` that never changes.
+We model our legendary item as a Java Record because these records are immutable by default, allowing us to hard-code the quality.
+
+Sealed interfaces help with the creation of a lightweight type hierarchy, allowing us to group all the items together for processing.
+
+```java
+public sealed interface PurchasableItem 
+  permits LegendaryItem, AgedBrie, CommonItem {
+  String name();
+  SellIn sellIn();
+  Quality quality();
+}
+
+public sealed interface LegendaryItem 
+  extends PurchasableItem 
+  permits Sulfuras {}
+
+public record AgedBrie(
+  UnboundedSellIn sellIn, 
+  BoundedQuality quality
+) implements MaturedCheese {
+  @Override public String name() { return "Aged Brie"; }
+}
+
+public record Sulfuras(
+  FixedSellIn sellIn
+) implements LegendaryItem {
+
+  private static final FixedQuality quality = FixedQuality.of(80);
+
+  @Override
+  public String name() { return "Sulfuras, Hand of Ragnaros"; }
+
+  @Override
+  public Quality quality() { return quality; }
+}
+
+// ...
+```
+
+By utilising Java's pattern matching and [record patterns](https://openjdk.org/jeps/440) (available in Java 21 and later), we can separate our code for processing purchasable items into distinct code paths.
+
+```java
+PurchasableItem item = // ...
+
+var updated = switch (item) {  
+  case CommonItem common when common.sellIn().isExpired() -> {}
+  case CommonItem common -> { /* ... */ }  
+  case Sulfuras(SellIn sellIn) -> { /* ... */ }  
+  case AgedBrie(SellIn sellIn, Quality quality) -> { /* ... */ }  
+}
+```
+
+Record patterns let you deconstruct records right in the patterns, giving you a more expressive and concise way to access record fields.
+Plus, the compiler makes sure you handle all the fields in the record pattern, which helps catch bugs that might pop up from incomplete data extraction.
```

Comment:
It might help to elaborate on this statement, do you mean that it forces the engineer to provide an implementation for each case?